src/main/java/com/couchbase/lite/replicator/PullerInternal.java

```diff
@@ -456,12 +462,22 @@ private void queueDownloadedRevision(RevisionInternal rev) {
             }
         }
 
-        if (rev != null && rev.getBody() != null)
-            rev.getBody().compact();
+        // NOTE: should not/not necessary to call Body.compact()
+        // new RevisionInternal(Map<string, Object>) creates Body instance only
+        // with `object`. Serializing object to json causes two unnecessary
+        // JSON serializations.
+
+        if (rev.getBody() != null)
+            queuedMemorySize.addAndGet(rev.getBody().getSize());
 
         downloadsToInsert.queueObject(rev);
-    }
 
+        // if queue memory size is more than maximum, force flush the queue.
+        if (queuedMemorySize.get() > MAX_QUEUE_MEMORY_SIZE) {
+            Log.d(TAG, "Flushing queued memory size at: " + queuedMemorySize);
+            downloadsToInsert.flushAll(true);
```

Comment:
queuedMemorySize is reset in Batcher.process() method if no items in queue.
```java
public void process(List<RevisionInternal> inbox) {
	insertDownloads(inbox);
	if (downloadsToInsert.count() == 0) {
		queuedMemorySize.set(0);
	}
}                
```
https://github.com/couchbase/couchbase-lite-java-core/pull/1577/files#diff-c739f8c5636953997d2d00324944fc7fR131